### PR TITLE
gh-120452: improve documentation about private name mangling

### PR DESCRIPTION
**Pull request by picnixz**
I do not have a NEWS entry but tell me if I need one.

(I'm shamefully pinging @erlend-aasland since you were quite reactive a few minutes ago!)

----
📚 Documentation preview 📚: 

- https://cpython-previews--120451.org.readthedocs.build/en/120451/reference/expressions.html#atom-identifiers
- https://cpython-previews--120451.org.readthedocs.build/en/120451/tutorial/classes.html#private-variables
- https://cpython-previews--120451.org.readthedocs.build/en/120451/faq/programming.html#i-try-to-use-spam-and-i-get-an-error-about-someclassname-spam

* Issue: gh-120452

### CODE REVIEW

**Comment by nineteendo** [comment]
Maybe it's worth linking to https://docs.python.org/3/tutorial/classes.html#private-variables?

**Comment by picnixz** [comment]
Oh great, there are actually TWO sections. Let me actually check whether I should rather link or not.

**Comment by picnixz** [comment]
Ok, there are some imprecision in the paragraph you just linked:

>  Any identifier of the form __spam (at least two leading underscores, at most one trailing underscore) is textually replaced with _classname__spam, where classname is the current class name with leading underscore(s) stripped.

And I think I know why I always forget about the "all underscore" case... Because it's not explained here.

File: Doc/faq/programming.rst [hunk 1]
```diff
@@ -1739,7 +1739,8 @@ Variable names with double leading underscores are "mangled" to provide a simple
 but effective way to define class private variables.  Any identifier of the form
 ``__spam`` (at least two leading underscores, at most one trailing underscore)
 is textually replaced with ``_classname__spam``, where ``classname`` is the
-current class name with any leading underscores stripped.
+current class name with any leading underscores stripped
+(see :ref:`here <private-name-mangling>` for details and special cases).
```
Comment (erlend-aasland) [inline Doc/faq/programming.rst]: Some comments:

1. In general it is good practice to make the link text meaningful on its own; this can be helpful for, for example, braille readers. IOW, _"here"_ won't pass 😉
2. Often, the `.. sealso::` directive is a good fit for cases like this.

**Comment by evle-zzz** [comment]
There are also weird situations with import statement, which, I belive, are not described by this. 
`import __super_module` is mangled like  `_A__super_module`
whereas `import __super_module.sub` remains not mangled.

Also, this explain classes named with underscores, but not identifiers within classes. Like, `self.__ = 1` within class with any name will not be mangled. 

Additionally it might be nice to explain the goal behind mangling and applicability, but that could be a more debatable topic that require longer discussion.

**Comment by picnixz** [comment]
> Also, this explain classes named with underscores, but not identifiers within classes. Like, self.__ = 1 within class with any name will not be mangled.

This one can indeed be important since I do not think it's ever been mentioned explicitly (like, the distinction between using the identifier via `self.` or `cls.` internally and externally).

> Additionally it might be nice to explain the goal behind mangling and applicability, but that could be a more debatable topic that require longer discussion.

This should be done in a separate PR IMO. There is already some kind of motivation where we justify its use for private/protected interfaces (not in those terms though).

>  import __super_module is mangled like _A__super_module
whereas import __super_module.sub remains not mangled.

Do you mean something like this:

```python
class A:
	import __hehe
	import __hehe.haha

A._A__hehe  # ok
A.__hehe    # not ok
A.__hehe.haha  # (maybe? I don't really know how it would look)
```

**Comment by picnixz** [comment]
Actually, this highlights the impossibility of importing the `__haha` module in the scope of a class *unless* the class only consists of underscores! The question I have now is: should I document this behaviour or not? I would put it in the *Expressions* section but not in the others because this is a quite convoluted case.

File: Doc/faq/programming.rst (hunk 1 above)
Comment (picnixz) [reply Doc/faq/programming.rst]: I wanted to avoid the `.. seealso::` because it would break the reading flow IMO (for this specific paragraph, not in general). On the other hand, the best title I could think of was `See the private identifier specifications` but I was not really sure whether it was the best or not.

I confess that I did not consider Braille readers...

**Comment by nineteendo** [comment]
> this highlights the impossibility of importing the `__haha` module

It's possible if you really need it:

```python
class A:
    __haha = __import__("__haha")
```

**Comment by picnixz** [comment]
I don't think I will document this hack (but I'll edit my answer) but thank you!

**Comment by picnixz** [comment]
FTR (and for future PRs), I'm force-pushing when I'm hotfixing typos in a commit (it does not make sense to review a commit with a known error).

**Comment by nineteendo** [comment]
Thanks, I always to prefer to know why a force push was necessary.

**Comment by picnixz** [comment]
In the end, I documented the hack because it feels a bit weird to say "you cannot do it" without telling you how to do it if you really need to. 

> Thanks, I always to prefer to know why a force push was necessary.

Yes, and I know that maintainers don't like force-pushing (and I don't like it either when working on Sphinx, it's hard to review and my "reviewed" changes becomes messy).

File: Doc/faq/programming.rst (hunk 1 above)
Comment (picnixz) [reply Doc/faq/programming.rst]: Actually, the `seealso` is fine visually so I'll use it more in the future:

image

I am a bit biased with the default themes in Sphinx where the See Also block is a very huge block...

**Comment by evle-zzz** [comment]
> There is already some kind of motivation where we justify its use for private/protected interfaces (not in those terms though).
There is, and it's misread and misused to often by newcomers, which is why it would benefit from rephrasing. But I agree, that's way wider topic that clarification that this PR provides. 

> Do you mean something like this:
Yes, like this.

> The question I have now is: should I document this behaviour or not?
There are actually three questions:
1. Is this an intended behaviour or just a fluke?
2. Is this part of language specification to be relied on or an implementation detail? It would be bad to document somehing that is not guaranteed. 
3. Should that be known by public?
Basically I'm ok with any wording if it's not wrong. Initial explanation from PR is not entirely correct, because real behaviour contradicts documentation. Just adding something like "Behaviour might be different for import statements, do not rely on mangling for that" or something like that could be a sufficient solution. 

BTW, Murphy from mCoding made a nice video with investigation on how mangling works in different scenarios. I might have missed some cases. Not sure if YouTube links are appropriate here, but should be searchable by "Every Python dev falls for this (name mangling)" name.

**Comment by nineteendo** [comment]
> 1. Is this an intended behaviour or just a fluke?
> 2. Is this part of language specification to be relied on or an implementation detail?

Yes, `__import__("__haha")` works the same as `getattr(self, "__haha")`, and I wouldn't expect anything different:

```python
class A:
    def __init__(self, haha) -> None:
        setattr(self, "__haha", haha)

    @property
    def haha(self):
        return getattr(self, "__haha")

print(A(123).haha) # 123
```

But that is also not documented here.

**Comment by picnixz** [comment]
> Is this an intended behaviour or just a fluke?

The way I see it still makes partially sense but this is subject to interpretation. When you do a `import __spam`, I assume that the parser first changes the name *and* tries to import it, which is correct according to the specs (transformation is done before code is emitted, this is what is being written). On the other hand, when you do `import __spam.util`, it feels to me that it's a bit inconsistent with what is being said. I would expect `__spam` being changed into `_Ham__spam` and then `import _Ham__spam.util` which should fail. 

@JelleZijlstra since you have quite a good knowledge of compiler-related stuff, do you have an explanation for the second behaviour?

> Is this part of language specification to be relied on or an implementation detail? It would be bad to document somehing that is not guaranteed.

I don't think it's an implementation detail. The only implementation detail is when the name is longer than 255 characters and its something that is in the docs already.

> Basically I'm ok with any wording if it's not wrong. Initial explanation from PR is not entirely correct

That's what I added in the expressions section (so the one that should serve as the formal definition).

> But that is also not documented here.

Actually, it's documented below in the Classes section (at the end of the paragraph). But I may move that one into the formal definition as well.

File: Doc/reference/expressions.rst [hunk 2]
```diff
@@ -83,17 +83,78 @@ exception.
...
-the identifier ``__spam`` occurring in a class named ``Ham`` will be transformed
-to ``_Ham__spam``.  This transformation is independent of the syntactical
-context in which the identifier is used.  If the transformed name is extremely
-long (longer than 255 characters), implementation defined truncation may happen.
-If the class name consists only of underscores, no transformation is done.
+.. rubric:: Private name mangling
+
+When an identifier that textually occurs in a class definition begins with two
+or more underscore characters and does not end in two or more underscores, it
+is considered a :dfn:`private name` of that class.
```
Comment (nineteendo) [inline Doc/reference/expressions.rst:91]: Maybe already mention this here?

```suggestion
is considered a :dfn:`private name` of that class. This doesn't make it
inaccessible from the outside, see below.
```

File: Doc/reference/expressions.rst [hunk 3]
```diff
@@ -83,17 +83,78 @@ exception.
...
+
+  For identifiers declared using :keyword:`import` statements, this rule is
+  slightly different. Indeed, importing a module with a private name directly
+  in a class body raises a :exc:`ModuleNotFoundError` (unless the class name
+  only consists of underscores), as illustrated by the following example:
+
+  .. code-block:: python
+
+     class Foo:
+         import __spam  # raises ModuleNotFoundError at runtime
```
Comment (nineteendo) [inline Doc/reference/expressions.rst]: Also note that this works if there's a module named `_Foo__spam`.

```suggestion
         import __spam  # tries to import _Foo__spam
```

Comment (picnixz) [reply Doc/reference/expressions.rst]: > Also note that this works if there's a module named _Foo__spam.

I feel that I need a separate section for importing private names, so that I can also put the file structure of the example. 

By the way, could you make your reviews in one batch instead of multiple comments (unless you have a specific reason not to)?

File: Doc/reference/expressions.rst (hunk 2 above)
Comment (picnixz) [reply Doc/reference/expressions.rst:91]: I think this is quite obvious otherwise it does not make sense to declare such identifier (unless I say "inaccessible from the outside" but then it would stretch the introductory paragraph about the private name a bit too much IMO).

**Comment by evle-zzz** [comment]
> > 1. Is this an intended behaviour or just a fluke?
> > 2. Is this part of language specification to be relied on or an implementation detail?
> 
> Yes, `__import__("__haha")` works the same as `getattr(self, "__haha")`, and I wouldn't expect anything different:

I'm talking about mangling rules as a whole, not `__import__` trick, this one is obvious.
I just doubt that anyone actually paid much attention to what would happen to import statements with double undrescores inside a class because of mangling to make this behaviour deliberate: it's an extremely weird scenario. As soon as it documented, we are forcing future parser changes to comply with this behaviour (or thinking about changing documentation again).



> On the other hand, when you do `import __spam.util`, it feels to me that it's a bit inconsistent with what is being said.

I have a hunch that in this case whole string is sent into mangling procedure and a dot prevents further actions silently. I'm too lazy to go ahead and check though. :-)

**Comment by JelleZijlstra** [comment]
I'll have to look at this later. @evle-zzz is probably right that the interaction with `import` may have been partly accidental, but backward compatibility probably means we can't change it now, so we may have to document and test the current behavior.

**Comment by picnixz** [comment]
Aaah there are sentences I forgot to end! Instead of force-pushing, just wait a bit before reviewing the latest changes then.

File: Doc/reference/expressions.rst [hunk 4]
```diff
@@ -83,18 +83,137 @@ exception.
...
+be performed via the :func:`__import__` function, in which case the usual
+transformation rule is applied:
+
+.. code-block:: python
+
+   class Bar:
+       # explicitly import '__bar' instead of '_Bar__bar'
+       __bar = __import__("__bar")
+   print(Bar._Bar__bar)  # <module '__bar' from '/__bar.py'>
+
```
Comment (nineteendo) [inline Doc/reference/expressions.rst]: No need to mention this for `Bar` too:

```suggestion
```

File: Doc/reference/expressions.rst [hunk 5]
```diff
@@ -83,18 +83,137 @@ exception.
...
+   a more complete example.
+
+.. _private-name-mangling-imports:
+
+.. rubric:: Mangled names in imports
+
+For identifiers declared using :keyword:`import` statements, the mangling
+rule is slightly different. Throughout this paragraph, assume that we have
+the following filesystem layout and, unless stated otherwise, the code snippets
+are in the ``__main__.py`` file.
```
Comment (nineteendo) [inline Doc/reference/expressions.rst]: No need to complicate this sentence when all snippets are in `__main__.py`.

```suggestion
the following filesystem layout and the code snippets are in the
``__main__.py`` file.
```

Comment (picnixz) [reply Doc/reference/expressions.rst]: Ah yes, I wasn't sure whether I'll need a special case but forgot to update the text.

File: Doc/reference/expressions.rst [hunk 6]
```diff
@@ -83,18 +83,131 @@ exception.
...
-long (longer than 255 characters), implementation defined truncation may happen.
-If the class name consists only of underscores, no transformation is done.
+Private name mangling
+^^^^^^^^^^^^^^^^^^^^^
 
+When an identifier that textually occurs in a class definition begins with two
+or more underscore characters and does not end in two or more underscores, it
+is considered a :dfn:`private name` of that class.
+
+.. seealso::
```
Comment (JelleZijlstra) [inline Doc/reference/expressions.rst:93]: I don't think the language reference should refer to the tutorial.

File: Doc/reference/expressions.rst [hunk 7]
```diff
@@ -83,18 +83,131 @@ exception.
...
+
+The transformation is independent of the syntactical context in which the
+identifier is used and the transformation rule is defined as follows:
+
+- If the class name consists only of underscores, the transformation is the
+  identity, e.g., the identifier ``__spam`` occurring in a class named ``_``
+  or ``__`` is left as is.
+
+- Otherwise, the transformation inserts the class name, with leading
+  underscores removed and a single underscore inserted, in front of
```
Comment (JelleZijlstra) [inline Doc/reference/expressions.rst]:
```suggestion
  underscores removed and a single leading underscore inserted, in front of
```

File: Doc/reference/expressions.rst [hunk 8]
```diff
@@ -83,18 +83,131 @@ exception.
...
+  the identifier, e.g., the identifier ``__spam`` occurring in a class
+  named ``Foo``, ``_Foo`` or ``__Foo`` is transformed to ``_Foo__spam``.
+
+.. _private-name-mangling-access:
+
+.. rubric:: Accessing members with mangled names
```
Comment (JelleZijlstra) [inline Doc/reference/expressions.rst]: I don't think the "rubric" directive is appropriate here; it is used almost exclusively for footnotes. Use a subheading instead.

File: Doc/reference/expressions.rst [hunk 9]
```diff
@@ -83,18 +83,131 @@ exception.
...
+
+The corresponding private member is accessed by its defining class using
```
Comment (JelleZijlstra) [inline Doc/reference/expressions.rst]: Why does this section need to be in the language reference? It doesn't seem to specify any behavior.

File: Doc/reference/expressions.rst [hunk 10]
```diff
@@ -83,18 +83,131 @@ exception.
...
+
+.. code-block:: python
+
+   class Foo:
+       import __pkg.mod
+
+   print(Foo._Foo__pkg)      # <module '__pkg' from '/__pkg/__init__.py'>
+   print(Foo._Foo__pkg.mod)  # <module '__pkg.mod' from '/__pkg/mod.py'>
+
+Note that a class whose name only consists of underscores does not
```
Comment (JelleZijlstra) [inline Doc/reference/expressions.rst]: This is implied by the specification of name mangling above.

File: Doc/reference/expressions.rst [hunk 11]
```diff
@@ -83,18 +83,131 @@ exception.
...
-removed and a single underscore inserted, in front of the name.  For example,
-the identifier ``__spam`` occurring in a class named ``Ham`` will be transformed
-to ``_Ham__spam``.  This transformation is independent of the syntactical
-context in which the identifier is used.  If the transformed name is extremely
-long (longer than 255 characters), implementation defined truncation may happen.
-If the class name consists only of underscores, no transformation is done.
+Private name mangling
+^^^^^^^^^^^^^^^^^^^^^
 
+When an identifier that textually occurs in a class definition begins with two
```
Comment (JelleZijlstra) [inline Doc/reference/expressions.rst:89]: I think the detailed discussion of imports below should be replaced with a more precise definition of what an "identifier" is here. It is not obvious that it includes attribute names and imported module names, as well as directly used names.

**Review by JelleZijlstra — CHANGES_REQUESTED** [review]
This goes into too much detail on the behavior around imports. The language reference should be more concise.

File: Doc/reference/expressions.rst (hunk 11 above)
Comment (JelleZijlstra) [reply Doc/reference/expressions.rst:89]: From looking at the implementation, the names that are mangled are:
* Any name used as the name of a variable that is assigned or read. However, the `__name__` attribute of nested functions, classes, and type aliases is not mangled.
* Any attribute accessed
* The name of imported modules and of names that are imported from a module, unless the module is part of a package (i.e., the name contains a dot).

**Comment by picnixz** [comment]
Thank you Jelle for your feedback! very much appreciated. I had the feelings that the section I expanded was indeed too long but I had various questions (I'll reply to your comments in one go so that it's easier for the discussion):

> I don't think the "rubric" directive is appropriate here; it is used almost exclusively for footnotes. Use a subheading instead.

Oh fine by me (I like using the `rubric` to create titles without indices but if you want to reseerve them for footnotes & co, it's fine).

> I don't think the language reference should refer to the tutorial.

In the identifier section, we talk about "classes" but those are not yet introduced (actually I put the wrong reference, it's not the tutorial but the compound statements on classes that should be referred to. However this still comes *after* the current section). 

> Why does this section need to be in the language reference? It doesn't seem to specify any behavior.

Now that I think about, it doesn't make sense to put it here; it's not a definition but rather a how-to-use. I will put it in the tutorial (and expand what is already said out there). By the way, this is unrelated, but the suggestion engine could actually suggest mangled names if the attribute we want to access is a private name. For instance, if you write `A.__a` instead of `A._A__` you get a `AttributeError: type object 'A' has no attribute '__a'`, but I think you could suggest `AttributeError: type object 'A' has no attribute '__a'. Did you mean: '_A__a'?`. For now, suggestion only pops if you type `_A__x` for instance, i.e., the `_A` prefix is needed.

> I think the detailed discussion of imports below should be replaced with a more precise definition of what an "identifier" is here. It is not obvious that it includes attribute names and imported module names, as well as directly used names.

Yes, and I'd move that discussion in a tutorial section instead (or I can delete it altogether, but I think it's nice to have examples). For me, the more examples you have, the easier it is for a beginner to understand what happens on such topics (especially if the topic is quite convoluted).

----

Before addressing the remaining comments (that I did not mark as resolved), I would like you to confirm whether the approach I described above would be fine.

**Comment by JelleZijlstra** [comment]
> > I don't think the language reference should refer to the tutorial.
> 
> In the identifier section, we talk about "classes" but those are not yet introduced (actually I put the wrong reference, it's not the tutorial but the compound statements on classes that should be referred to. However this still comes _after_ the current section).

Linking to other sections in the language reference is fine.
 
> > Why does this section need to be in the language reference? It doesn't seem to specify any behavior.
> 
> Now that I think about, it doesn't make sense to put it here; it's not a definition but rather a how-to-use. I will put it in the tutorial (and expand what is already said out there). By the way, this is unrelated, but the suggestion engine could actually suggest mangled names if the attribute we want to access is a private name. For instance, if you write `A.__a` instead of `A._A__` you get a `AttributeError: type object 'A' has no attribute '__a'`, but I think you could suggest `AttributeError: type object 'A' has no attribute '__a'. Did you mean: '_A__a'?`. For now, suggestion only pops if you type `_A__x` for instance, i.e., the `_A` prefix is needed.

I don't think this needs to be in the tutorial either. The tutorial is meant to introduce people to the language, and I think very few Python programmers will ever run into the interaction between import statements and name mangling.

> > I think the detailed discussion of imports below should be replaced with a more precise definition of what an "identifier" is here. It is not obvious that it includes attribute names and imported module names, as well as directly used names.
> 
> Yes, and I'd move that discussion in a tutorial section instead (or I can delete it altogether, but I think it's nice to have examples). For me, the more examples you have, the easier it is for a beginner to understand what happens on such topics (especially if the topic is quite convoluted).

Some examples could be useful but the section should be kept concise. The language reference is a reference for those who want to understand the language's behavior, not an introduction for new programmers.

**Comment by picnixz** [comment]
> I don't think this needs to be in the tutorial either

Actually, it was about the whole "how to access mangled names". I can remove the examples on the import though but will talk about mangling for regular attributes.

So here's what I suggest:

- Remove the import examples, replacing them by a more precise definition of an identifier.
- Move the 'Accessing private names' into the classes tutorial (there's already something out there actually, but I may rephrase it instead of having a huge block of text?). Maybe I won't even need this one.

**Comment by JelleZijlstra** [comment]
Makes sense, accessing mangled attributes is a common enough need that it does make sense to say something about it in the tutorial.

**Comment by picnixz** [comment]
Jelle, I have made the requested changes; please review again. I've decided to put the example on accessing the name in the FAQ because of the question `I try to use __spam and I get an error about _SomeClassName__spam`.

File: Doc/reference/expressions.rst [hunk 12]
```diff
@@ -92,15 +92,29 @@ is considered a :dfn:`private name` of that class.
...
+identifier is used but only the following private identifiers are mangled:
+
+- The name of imported modules, e.g., ``__spam`` in ``import __spam``.
+  If the module is part of a package (i.e., its name contains a dot),
+  the name is *not* mangled, e.g., the ``__foo`` in ``import __foo.bar``
+  is not mangled.
+
+- The name of an imported member, e.g., ``__f`` in ``from spam import __f``.
+
+- Any name used as the name of a variable that is assigned or read or any
```
Comment (JelleZijlstra) [inline Doc/reference/expressions.rst]: Put this first since it's far more important than the import-related ones.

**Comment by picnixz** [comment]
@JelleZijlstra Friendly ping for reviewing this one again (or drop the PR)

File: Doc/reference/expressions.rst [hunk 13]
```diff
@@ -83,18 +83,47 @@ exception.
...
+The transformation rule is defined as follows:
+
+- If the class name consists only of underscores, the transformation is the
+  identity, e.g., the identifier ``__spam`` occurring in a class named ``_``
+  or ``__`` is left as is.
+
+- Otherwise, the transformation inserts the class name, with leading
+  underscores removed and a single leading underscore inserted, in front
+  of the identifier, e.g., the identifier ``__spam`` occurring in a class
+  named ``Foo``, ``_Foo`` or ``__Foo`` is transformed to ``_Foo__spam``.
```
Comment (JelleZijlstra) [inline Doc/reference/expressions.rst]: Let's swap these two paragraphs: put the common case first, and leave the edge case (a rather extreme edge case) last.
```suggestion
- The class name, with leading underscores removed and a single leading
  underscore inserted, is inserted in front of the identifier, e.g., the
  identifier ``__spam`` occurring in a class named ``Foo``, ``_Foo`` or
  ``__Foo`` is transformed to ``_Foo__spam``.
  
- If the class name consists only of underscores, the transformation is the
  identity, e.g., the identifier ``__spam`` occurring in a class named ``_``
  or ``__`` is left as is.
```

File: Doc/faq/programming.rst [hunk 14]
```diff
@@ -1741,11 +1741,34 @@ but effective way to define class private variables.  Any identifier of the form
 is textually replaced with ``_classname__spam``, where ``classname`` is the
 current class name with any leading underscores stripped.
 
+More generally, the corresponding private member is accessed by its defining
```
Comment (JelleZijlstra) [inline Doc/faq/programming.rst]: Do we need this extra paragraph? I'm conscious that when changing docs, we're always likely to feel that more documentation is needed, leading to the docs becoming more and more verbose over time. This FAQ entry is slightly imprecise relative to the language reference, but that's fine; it's not clear to me that the extra paragraph is necessary.

Comment (picnixz) [reply Doc/faq/programming.rst]: Mmh, maybe not but I think the example should be kept (it's a FAQ and generally examples in such cases are much more easy to understand). I admit that the paragraph could be more confusing than helping.

File: Doc/reference/expressions.rst (hunk 13 above)
Comment (picnixz) [reply Doc/reference/expressions.rst]: I'm sorry that I actually changed the order of the other but forgot that one.. Thanks Jelle!

File: Doc/faq/programming.rst (hunk 14 above)
Comment (JelleZijlstra) [reply Doc/faq/programming.rst]: How about replacing this paragraph with "The identifier can be used unchanged within the class, but to access it outside the class, the mangled name must be used:", followed by the example?

Comment (picnixz) [reply Doc/faq/programming.rst]: Oh sure. Actually, I wanted to entirely suppress the paragraph and just show the example, but I can add this introductory sentence!

**Review by JelleZijlstra — APPROVED** [review]

File: Doc/faq/programming.rst [hunk 15]
```diff
@@ -1741,11 +1741,32 @@ but effective way to define class private variables.  Any identifier of the form
...
+           return 2 * self.__one()
+
+   class B(A):
+       def three(self):
+           return 3 * self._A__one()
+
+   four = 4 * A()._A__one()
+
 This doesn't guarantee privacy: an outside user can still deliberately access
 the "_classname__spam" attribute, and private values are visible in the object's
```
Comment (nineteendo) [inline Doc/faq/programming.rst]: Isn't this a bit redundant now?

Comment (picnixz) [reply Doc/faq/programming.rst]: Mmmh... that's right. How about the following paragraph:

```rst
The identifier can be used unchanged within the class, but to access it outside
the class, the mangled name must be used:

<EXAMPLE>

In particular, this does not guarantee privacy since an outside user can still
deliberately access the private attribute, and private values remain visible in
the object's ``__dict__``.  Many Python programmers never bother to use private
variable names at all.
```

**OR**

```diff
-In particular, this does not guarantee privacy since an outside user can still
-deliberately access the private attribute, and private values remain visible in
-the object's ``__dict__``.  Many Python programmers never bother to use private
-variable names at all.
+In particular, this does not guarantee privacy since an outside user can still
+deliberately access the private attribute; many Python programmers never bother
+to use private variable names at all.
```

@JelleZijlstra

Comment (JelleZijlstra) [reply Doc/faq/programming.rst]: sounds good!

Comment (picnixz) [reply Doc/faq/programming.rst]: First or second suggestion? (I think the second is easier to digest for a beginner)

Comment (JelleZijlstra) [reply Doc/faq/programming.rst]: I like the second as it is more concise.

**Comment by picnixz** [comment]
The journey was long but I think we eventually managed to effictively improve the documentation. And I'd bet that I'll forget about it next time I'll face such issue :D
